accepted/2021/experimental-features/experimental-features.md

```diff
@@ -0,0 +1,390 @@
+    <TargetFramework>net6.0</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <EnableExperimentalFeature Include="StaticInterfaceMembers" />
+  </ItemGroup>
+
+</Project>
+```
+
+### Assembly info generation
+
+When we generate the `AssemblyInfo.cs`, we'll include all experimental features
+as an assembly level attribute (the attribute and the meaning is discussed
+[later in this document](#api-analyzer)).
+
+For example, this project information:
+
+```xml
+<Project>
+
+  <ItemGroup>
+    <EnableExperimentalFeature Include="StaticInterfaceMembers" />
+    <EnableExperimentalFeature Include="FancyGC" />
+  </ItemGroup>
+
+</Project>
+```
+
+Would result in these assembly-level attributes:
+
+```C#
+[assembly: ExperimentalFeature("StaticInterfaceMembers")]
+[assembly: ExperimentalFeature("FancyGC")]
+```
+
+This means you can look at a binary and tell which experimental features were
+used to build it. This can be used for auditing and for detecting transitive
+dependencies ([discussed later](#api-analyzer)).
+
+### runtime.json
+
+The `ExperimentalFeature` should be written into the `runtime.json` file such
+that the host/runtime can use it to make decisions.
+
+### Compilation context
+
+The `ExperimentalFeature` needs to be passed to the compilers. Roslyn (C#/VB)
+already has a mechanism for this which is exposed to both the compiler as well
+as for analyzers/source generators. Alternatively, we could say the information
+about experimental features can be retrieved from the generated assembly-level
+attributes.
+
+***Note:*** F# doesn't have this capability today.
+
+### API Analyzer
+
+We'll add an attribute that we can put on APIs we consider experimental. In some
+cases it might not be possible to mark an API (for example, because it's a new
+type system feature) in which case the feature would either be guarded by new
+language syntax or by a custom analyzer.
+
+```C#
+namespace System.Runtime.Versioning
+{
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Module |
+                    AttributeTargets.Class |
+                    AttributeTargets.Struct |
+                    AttributeTargets.Enum |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Method |
+                    AttributeTargets.Property |
+                    AttributeTargets.Field |
+                    AttributeTargets.Event, AllowMultiple = true, Inherited = false)]
+    public sealed class ExperimentalFeatureAttribute : Attribute
+    {
+        public ExperimentalFeatureAttribute(string featureName);
+        public string FeatureName { get; }
+    }
+}
+```
+
+We'll ship a built-in analyzer that will report diagnostics when an API is being
+called that is marked with this attribute and the feature name is not contained
+in the project's `ExperimentalFeature` items. The default severity is `error`.
+
+This analyzer will also validate that all referenced assemblies only have
```

Comment:
I think the analysis should be finer-grain. The parts of the conuming project that consume parts of a reference assembly that is in "preview" should themselves be marked as in "preview".

Let's say we add `INumeric` to the BCL as "preview" feature and some upper layers of the BCL expose more types or APIs on top of it. Those specific upper parts should be marked as "preview" (no less).